File: pkg/entdbadapter/hooks.go
package entdbadapter

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/fastschema/fastschema/db"
	"github.com/fastschema/fastschema/schema"
)

func runPreDBQueryHooks(ctx context.Context, client db.Client, option *db.QueryOption) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PreDBQuery) > 0 {
		for _, hook := range hooks.PreDBQuery {
			if err := hook(ctx, option); err != nil {
				return fmt.Errorf("pre query hook: %w", err)
			}
		}
	}

	return nil
}

func runPostDBQueryHooks(
	ctx context.Context,
	client db.Client,
	option *db.QueryOption,
	entities []*schema.Entity,
) (_ []*schema.Entity, err error) {
	if client == nil {
		return entities, nil
	}

	hooks := client.Hooks()
	if len(hooks.PostDBQuery) > 0 {
		for _, hook := range hooks.PostDBQuery {
			entities, err = hook(ctx, option, entities)
			if err != nil {
				return nil, fmt.Errorf("post query hook: %w", err)
			}
		}
	}

	return entities, nil
}

func runPreDBExecHooks(ctx context.Context, client db.Client, option *db.QueryOption) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PreDBExec) > 0 {
		for _, hook := range hooks.PreDBExec {
			if err := hook(ctx, option); err != nil {
				return fmt.Errorf("pre exec hook: %w", err)
			}
		}
	}

	return nil
}

func runPostDBExecHooks(
	ctx context.Context,
	client db.Client,
	option *db.QueryOption,
	result sql.Result,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PostDBExec) > 0 {
		for _, hook := range hooks.PostDBExec {
			if err := hook(ctx, option, result); err != nil {
				return fmt.Errorf("post exec hook: %w", err)
			}
		}
	}

	return nil
}

func runPreDBCreateHooks(
	ctx context.Context,
	client db.Client,
	schema *schema.Schema,
	createData *schema.Entity,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PreDBCreate) > 0 {
		for _, hook := range hooks.PreDBCreate {
			if err := hook(ctx, schema, createData); err != nil {
				return fmt.Errorf("pre create hook: %w", err)
			}
		}
	}

	return nil
}

func runPostDBCreateHooks(
	ctx context.Context,
	client db.Client,
	schema *schema.Schema,
	createData *schema.Entity,
	createdID uint64,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PostDBCreate) > 0 {
		for _, hook := range hooks.PostDBCreate {
			if err := hook(ctx, schema, createData, createdID); err != nil {
				return fmt.Errorf("post create hook: %w", err)
			}
		}
	}

	return nil
}

func runPreDBUpdateHooks(
	ctx context.Context,
	client db.Client,
	schema *schema.Schema,
	predicates []*db.Predicate,
	updateData *schema.Entity,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PreDBUpdate) > 0 {
		for _, hook := range hooks.PreDBUpdate {
			if err := hook(ctx, schema, predicates, updateData); err != nil {
				return fmt.Errorf("pre update hook: %w", err)
			}
		}
	}

	return nil
}

func runPostDBUpdateHooks(
	ctx context.Context,
	client db.Client,
	schema *schema.Schema,
	predicates []*db.Predicate,
	updateData *schema.Entity,
	originalEntities []*schema.Entity,
	affected int,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PostDBUpdate) > 0 {
		for _, hook := range hooks.PostDBUpdate {
			if err := hook(ctx, schema, predicates, updateData, originalEntities, affected); err != nil {
				return fmt.Errorf("post update hook: %w", err)
			}
		}
	}

	return nil
}

func runPreDBDeleteHooks(
	ctx context.Context,
	client db.Client,
	schema *schema.Schema,
	predicates []*db.Predicate,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PreDBDelete) > 0 {
		for _, hook := range hooks.PreDBDelete {
			if err := hook(ctx, schema, predicates); err != nil {
				return fmt.Errorf("pre delete hook: %w", err)
			}
		}
	}

	return nil
}

func runPostDBDeleteHooks(
	ctx context.Context,
	client db.Client,
	schema *schema.Schema,
	predicates []*db.Predicate,
	originalEntities []*schema.Entity,
	affected int,
) error {
	if client == nil {
		return nil
	}

	hooks := client.Hooks()
	if len(hooks.PostDBDelete) > 0 {
		for _, hook := range hooks.PostDBDelete {
			if err := hook(ctx, schema, predicates, originalEntities, affected); err != nil {
				return fmt.Errorf("post delete hook: %w", err)
			}
		}
	}

	return nil
}

File: examples/plugins/data/plugins/hello/hooks.js
'use strict';

/**
 * @param {FsContext} ctx
 * @param {FsQueryOption} option
 */
export const preDBQuery = (ctx, option) => {
  console.log('[plugin] PreDBQuery this runs before the query is executed');
};

/**
 * @param {FsContext} ctx
 * @param {FsQueryOption} option
 * @param {FsEntity[]} entities
 */
export const postDBQuery = (ctx, option, entities) => {
  console.log(
    '[plugin] PostDBQuery this runs after the query is executed, result: ',
    JSON.stringify(entities.map(e => e.ToMap()), null, 2)
  );
};

/**
 * @param {FsContext} ctx
 * @param {FsQueryOption} option
 */
export const preDBExec = (ctx, option) => {
  console.log('[plugin] PreDBExec this runs before the exec is executed');
};

/**
 * @param {FsContext} ctx
 * @param {FsQueryOption} option
 * @param {FsDbExecResult} result
 */
export const postDBExec = (ctx, option, result) => {
  console.log('[plugin] PostDBExec this runs after the exec is executed');
  console.log('insertId:', result.LastInsertId());
  console.log('affectedRows:', result.RowsAffected());
};

/**
 * @param {FsContext} ctx
 * @param {SchemaRawData} schema
 * @param {FsEntity} createData
 */
export const preDBCreate = (ctx, schema, createData) => {
  console.log('[plugin] PreDBCreate this runs before the create is executed');
  console.log('schema:', schema);
  console.log('createData:', createData.ToMap());
};

/**
 * @param {FsContext} ctx
 * @param {SchemaRawData} schema
 * @param {FsEntity} createData
 * @param {number} id
 */
export const postDBCreate = (ctx, schema, createData, id) => {
  console.log('[plugin] PostDBCreate this runs after the create is executed');
  console.log('id:', id);
  console.log('createData:', createData.ToMap());
};

/**
 * @param {FsContext} ctx
 * @param {SchemaRawData} schema
 * @param {FsDbPredicate[]} predicates
 * @param {FsEntity} updateData
 */
export const preDBUpdate = (ctx, schema, predicates, updateData) => {
  console.log('[plugin] PreDBUpdate this runs before the update is executed');
  console.log('schema:', schema);
  console.log('id:', predicates);
  console.log('createData:', updateData.ToMap());
};

/**
 * @param {FsContext} ctx
 * @param {SchemaRawData} schema
 * @param {FsDbPredicate[]} predicates
 * @param {FsEntity} updateData
 * @param {FsEntity[]} originalEntities
 * @param {number} affected
 */
export const postDBUpdate = (
  ctx,
  schema,
  predicates,
  updateData,
  originalEntities,
  affected
) => {
  console.log('[plugin] PostDBUpdate this runs after the update is executed');
  console.log('schema:', schema);
  console.log('predicates:', predicates);
  console.log('updateData:', updateData);
  console.log('originalEntities:', originalEntities.map(e => e.ToMap()));
  console.log('affected:', affected);
};

/**
 * @param {FsContext} ctx
 * @param {SchemaRawData} schema
 * @param {FsDbPredicate[]} predicates
 */
export const preDBDelete = (ctx, schema, predicates) => {
  console.log('[plugin] PreDBDelete this runs before the delete is executed');
  console.log('schema:', schema);
  console.log('predicates:', predicates);
};

/**
 * @param {FsContext} ctx
 * @param {SchemaRawData} schema
 * @param {FsDbPredicate[]} predicates
 * @param {FsEntity[]} originalEntities
 * @param {number} affected
 */
export const postDBDelete = (
  ctx,
  schema,
  predicates,
  originalEntities,
  affected
) => {
  console.log('[plugin] PostDBDelete this runs after the delete is executed');
  console.log('schema:', schema);
  console.log('predicates:', predicates);
  console.log('originalEntities:', originalEntities);
  console.log('affected:', affected);
};

/** @param {FsContext} ctx */
export const preResolve = (ctx) => {
  console.log('[plugin] PreResolve this runs before the request is resolved');
};

/** @param {FsContext} ctx */
export const postResolve = (ctx) => {
  console.log('[plugin] PostResolve this runs after the request is resolved');
};

export default {
  preResolve,
  postResolve,

  preDBQuery,
  postDBQuery,

  preDBExec,
  postDBExec,

  preDBCreate,
  postDBCreate,

  preDBUpdate,
  postDBUpdate,

  preDBDelete,
  postDBDelete,
};
